### PR TITLE
Make scanner launch Tor for itself

### LINKED ISSUE #77: Have sbws client launch tor itself
Stem can do it. https://stem.torproject.org/api/process.html

But I'm not sure we should. In fact, while it sounds cool, I'm leaning towards "don't do this."

Advantages:

- Most/all of the Tor options for sbws client are no longer necessary. sbws can configure Tor exactly how it wants. In fact, it can tell Tor to pick whatever it wants for a SocksPort, and sbws can use the ControlPort to figure out what was chosen)

- On the client side, the user doesn't even need to think about Tor as long as they have it installed.

- If sbws ever happened to need a custom Tor or a newer Tor than is commonly found in package managers (which I really hope is never the case), then implementing this the right way allows for the user to specify a custom path to their desired Tor

Disadvantages/unanswered questions:

- It's unclear what should happen for commands that need Tor but don't run continuously. Right now there are not any, but `sbws generate` might in the future. Should they really have to spend some time bootstrapping Tor just to determine the value of a consensus parameter? 

- What should happen if `sbws client` is running, using its Tor process, and we then run some other sbws command that requires Tor? Should another Tor process be started? Should it detect that an "sbws-owned Tor" already exists and use it?

Random thoughts about implementation details:

- Put the Tor data directory in `~/.sbws/tordata`. Could also put it in a fresh temporary directory every run, but I'm not sure that's a good idea.

- By default sbws should look for `tor` in the user's `$PATH`. But if some config option is set (say for example `tor_bin` in the section `[tor]`), it should use that instead.

- If Sbws cannot find Tor, it should only fail hard when running a command that requires it

- It's probably possible to support both "use the system Tor that's already running" and "launch our own Tor" cases, but for simplicity's sake, I think we should only support one or the other.

### PR DESCRIPTION
**Pull request by pastly**
Merging this closes #77

Note how most of the options in the `[tor]` section will go away. Your `[tor]` section in the config will most likely be empty now. @juga0, I highlight you because AFAIK you're the only other person running sbws occasionally.

### CODE REVIEW

File: tests/testnets/simple.common/01-gen-configs.sh [hunk 1]
```diff
@@ -149,10 +148,12 @@ datadir = \${sbws_home}/datadir
 sbws_home = $(pwd)
 
 [tor]
-control_type = socket
-control_location = \${paths:sbws_home}/$scanner_tor_socks_proxy_nick/control_socket
-socks_host = $scanner_tor_socks_proxy_ip
-socks_port = $socksport
+extra_lines =
+    TestingTorNetwork 1
+    NumCPUs 1
```
Comment (teor2345) [inline tests/testnets/simple.common/01-gen-configs.sh]: Why NumCPUs 1?
Seems like an odd option.

File: tests/testnets/simple.common/01-gen-configs.sh [hunk 2]
```diff
@@ -149,10 +148,12 @@ datadir = \${sbws_home}/datadir
...
+    LogTimeGranularity 1
+    SafeLogging 0
```
Comment (teor2345) [inline tests/testnets/simple.common/01-gen-configs.sh]: Consider UseEntryGuards 0 to disable pathbias warnings.
See #146.

File: tests/testnets/simple.common/01-gen-configs.sh (hunk 1 above)
Comment (pastly) [reply tests/testnets/simple.common/01-gen-configs.sh]: There's ~12 Tor processes that are run at once in these tests and I thought I would attempt to not peg all of everyone's CPU cores, if possible.

File: tests/testnets/simple.common/01-gen-configs.sh (hunk 2 above)
Comment (pastly) [reply tests/testnets/simple.common/01-gen-configs.sh]: @teor2345 [I don't think we should use FetchDirInfo*Early](https://github.com/pastly/simple-bw-scanner/issues/146#issuecomment-385976197)

Also, notice the inclusion of them in initial commits on this branch is undone by latter commits (ce2187e).

Comment (teor2345) [reply tests/testnets/simple.common/01-gen-configs.sh]: I did not notice, I was using the files view.

Comment (pastly) [reply tests/testnets/simple.common/01-gen-configs.sh]: @juga0 these lines are *in addition* to the torrc config lines you linked to. See how in [b378bf63](https://github.com/pastly/simple-bw-scanner/pull/147/commits/b378bf63e09fcbc3a060c1d387363975f89e29b9#diff-a250492b468dd26963e8a1c2343cb1e1R175) I added the ability for torrc options to be read from config.ini.

So in the test networks, all the options in `sbws/util/stem.py launch_tor()` are set, and then these `extra_lines` are added.

Does that make sense? I'm not sure I've addressed your concern.

**Comment by pastly** [comment]
@juga, fcac281 addresses your suggestion on IRC to move as many torrc options as possible to a global variable.

I looked at using chutney when writing the stuff in `tests/testnets/`. Here are some of the reasons I didn't go with chutney

- It's another dependency
- It's harder to control the chutney version the user is using. It's easier to just have the test network generation scripts source-controlled right with the sbws code.
- Chutney is complex, especially in the torrc templating. It's complex because it wants to be flexible. I don't need flexible. I just want reproducible. I thought I could implement only the things I need in a couple of scripts and have those scripts be significantly easier to grok than chutney. IMHO, I succeeded. (But I also wrote them, so I'm heavily biased)

----

@juga0, feel like this is merge ready now?

**Comment by juga0** [comment]
This branch was already merged.
Re. chutney, realized there isn't doc for testnets (i know i could look code...).
Is it possible right now to connect to the a running Tor (initiated by testnets) to run a unit test?

**Comment by juga0** [comment]
Re Chutney dependency and versions: in bwscanner it was only being retrieved automatically by travis from git to run tests, so normal installation would not install/make use of chutney.

**Comment by juga0** [comment]
>> Is it possible right now to connect to the a running Tor (initiated by testnets) to run a unit test?
> sbws scanner first tries to connect to a running Tor.
yeah, i meant unit tests, not the scanner. Rephrasing:
"Is it possible a unit test connect to a running Tor"?
> I'm working on figuring out a good way to have the unit tests launch Tor for testing. For example, each test should not necessarily have to launch Tor for itself because that would take a very long time.
Exactly, i said "connect to", and i noted that launching Tor would be slow and not reliable in: https://github.com/pastly/simple-bw-scanner/pull/174/commits/832e33902c0ecc445629a23e743c8d6760cdc435#diff-814ac08b96f7b5924db4cc9038327f11R5

I found this very easy to do with chutney

**Pull request merged**